devicetypes/stelpro/stelpro-ki-thermostat.src/stelpro-ki-thermostat.groovy

```diff
@@ -0,0 +1,515 @@
+			])
+	}
+
+	log.debug "Parse returned $result"
+	result
+}
+
+def updateWeather() {
+	log.debug "updating weather"
+	def weather
+	// If there is a zipcode defined, weather forecast will be sent. Otherwise, no weather forecast.
+	if (settings.zipcode) {
+		log.debug "ZipCode: ${settings.zipcode}"
+		weather = getWeatherFeature("conditions", settings.zipcode)
+
+		// Check if the variable is populated, otherwise return.
+		if (!weather) {
+			log.debug("Something went wrong, no data found.")
+			return false
+		}
+
+		def locationScale = getTemperatureScale()
+		def tempToSend = (locationScale == "C") ? weather.current_observation.temp_c : weather.current_observation.temp_f
+		log.debug("Outdoor Temperature: ${tempToSend} ${locationScale}")
+		// Right now this can disrupt device health if the device is
+		// currently offline -- it would be erroneously marked online.
+		//sendEvent(name: 'outsideTemp', value: tempToSend)
+		setOutdoorTemperature(tempToSend)
+	}
+}
+
+def scheduledUpdateWeather() {
+	def actions = updateWeather()
+
+	if (actions) {
+		sendHubCommand(actions)
+	}
+}
+
+// Command Implementations
+
+/**
+ * PING is used by Device-Watch in attempt to reach the Device
+ **/
+def ping() {
+	log.debug "ping()"
+	zwave.sensorMultilevelV3.sensorMultilevelGet().format()
+}
+
+def poll() {
+	log.debug "poll()"
+	delayBetween([
+			updateWeather(),
+			zwave.thermostatOperatingStateV1.thermostatOperatingStateGet().format(),
+			zwave.thermostatModeV2.thermostatModeGet().format(),
+			zwave.thermostatSetpointV2.thermostatSetpointGet(setpointType: 1).format(),
+			zwave.sensorMultilevelV3.sensorMultilevelGet().format() // current temperature
+		], 100)
+}
+
+// Event Generation
+def zwaveEvent(thermostatsetpointv2.ThermostatSetpointReport cmd) {
+	def cmdScale = cmd.scale == 1 ? "F" : "C"
+	def temp;
+	float tempfloat;
+	def map = [:]
+
+	if (cmd.scaledValue >= 327 ||
+		cmd.setpointType != thermostatsetpointv2.ThermostatSetpointReport.SETPOINT_TYPE_HEATING_1) {
+		return [:]
+	}
+	temp = convertTemperatureIfNeeded(cmd.scaledValue, cmdScale, cmd.precision)
+	tempfloat = (Math.round(temp.toFloat() * 2)) / 2
+	map.value = tempfloat
+
+	map.unit = getTemperatureScale()
+	map.displayed = false
+	map.name = "heatingSetpoint"
+	map.data = [heatingSetpointRange: heatingSetpointRange]
+
+	// So we can respond with same format
+	state.size = cmd.size
+	state.scale = cmd.scale
+	state.precision = cmd.precision
+
+	map
+}
+
+def zwaveEvent(sensormultilevelv3.SensorMultilevelReport cmd) {
+	def temp
+	float tempfloat
+	def format
+	def map = [:]
+
+	if (cmd.sensorType == sensormultilevelv3.SensorMultilevelReport.SENSOR_TYPE_TEMPERATURE_VERSION_1) {
+		map.value = convertTemperatureIfNeeded(cmd.scaledSensorValue, cmd.scale == 1 ? "F" : "C", cmd.precision)
+		map.unit = getTemperatureScale()
+		map.name = "temperature"
+
+		temp = map.value
+		// The specific values checked below represent ambient temperature alarm indicators
```

Comment:
I wonder now about the wisdom of handling these special cases _after_ the conversion instead of just comparing against the raw hex, but it's fine